stimela/kitchen/recipe.py

```diff
@@ -356,18 +365,81 @@ def __post_init__ (self):
                 steps[label] = Step(**stepconfig)
             self.steps = steps
         # check that assignments don't clash with i/o parameters
-        for assign, assign_label in [(self.assign, "assign")] + [(step.assign, f"{label}.assign") for label, step in self.steps.items()]:
-            for key in assign:
-                for io, io_label in [(self.inputs, "inputs"), (self.outputs, "outputs")]:
-                    if key in io:
-                        raise RecipeValidationError(f"'{assign_label}.{key}' clashes with recipe {io_label}")
+        self.validate_assignments(self.assign, self.assign_based_on, self.name)
+
         # check that for-loop variable does not clash
         if self.for_loop:
             for io, io_label in [(self.inputs, "inputs"), (self.outputs, "outputs")]:
                 if self.for_loop.var in io:
                     raise RecipeValidationError(f"'for_loop.var={self.for_loop.var}' clashes with recipe {io_label}")
         # map of aliases
         self._alias_map = None
+        # set of keys protected from assignment
+        self._protected_from_assign = set()
+
+    def protect_from_assignments(self, keys):
+        self._protected_from_assign.update(keys)
+        #self.log.debug(f"protected from assignment: {self._protected_from_assign}")
+
+    def validate_assignments(self, assign, assign_based_on, location):
+        # collect a list of all assignments
+        assignments = OrderedDict()
+        for key in assign:
+            assignments[key] = "assign"
+        for basevar, lookup_list in assign_based_on.items():
+            if not isinstance(lookup_list, Mapping):
+                raise RecipeValidationError(f"{location}.{assign_based_on}.{basevar}: mapping expected")
+            # for assign_list in lookup_list.values():
```

Comment:
Why not delete this old code. We have the git history if you need it again in the future